lang/cs/Org.Apache.REEF.Network/Elastic/Task/Workflow.cs

```diff
@@ -0,0 +1,298 @@
+    /// Exception rised during execution are managed by the framework and recovered through the user-defined
+    /// policies / mechanisms.
+    /// </summary>
+    [Unstable("0.16", "API may change")]
+    public sealed class Workflow : IEnumerator<IElasticOperator>
+    {
+        private static readonly Logger LOGGER = Logger.GetLogger(typeof(Workflow));
+
+        private int _position = -1;
+        private bool _failed;
+        private bool _disposed;
+        private List<int> _iteratorsPosition;
+
+        private readonly object _lock;
+        private readonly IList<IElasticOperator> _operators;
+        private readonly CancellationSource _cancellationSource;
+        private readonly bool _isRescheduled;
+
+        /// <summary>
+        /// Injectable constructor.
+        /// </summary>
+        /// <param name="cancellationSource"></param>
+        [Inject]
+        private Workflow(
+            [Parameter(typeof(GroupCommunicationConfigurationOptions.IsRescheduled))] bool isRescheduled,
+            CancellationSource cancellationSource)
+        {
+            _operators = new List<IElasticOperator>();
+            _failed = false;
+            _disposed = false;
+            _lock = new object();
+            _iteratorsPosition = new List<int>();
+            _cancellationSource = cancellationSource;
+            _isRescheduled = isRescheduled;
+        }
+
+        /// <summary>
+        /// The current iteration value.
+        /// </summary>
+        public object Iteration
+        {
+            get
+            {
+                if (_iteratorsPosition.Count == 0)
+                {
+                    return 0;
+                }
+                else
+                {
+                    var iterPos = _iteratorsPosition[0];
+                    var iterator = _operators[iterPos] as IElasticIterator;
+                    return iterator.Current;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Try to move to the next operation in the workflow.
+        /// </summary>
+        /// <returns></returns>
+        public bool MoveNext()
+        {
+            _position++;
+
+            if (_failed || _cancellationSource.IsCancelled)
+            {
+                return false;
+            }
+
+            // Check if we need to iterate
+            if (_iteratorsPosition.Count > 0 && _position == _iteratorsPosition[0])
+            {
+                var iteratorOperator = _operators[_position] as IElasticIterator;
+
+                if (iteratorOperator.MoveNext())
+                {
+                    _position++;
+                    ResetOperatorPositions();
+
+                    return true;
+                }
+                else
+                {
+                    if (_iteratorsPosition.Count > 1)
+                    {
+                        _iteratorsPosition.RemoveAt(0);
```

Comment:
this operation can be expensive. Looks like we always add to the end of the list and peek/remove at 0. Maybe use `LinkedList` or `Queue` instead?

Reply:
Mmm not sure:
1. _iteratorsPosition is actually very small because it contains the number of iterator operators of the workflow. 99% of pipelines / workflows will have 1 iterator. I saw distributed trees implementations with 2 iterators, but this is probably the worst case.
2. addition and removal happen once (the former at the init of the workflows, the latter when each iterator is exhausted)
3. count is used more often and I think that list (which is backed by an array) should be faster the linkedlist and queues (although I am not sure how these are implemented, i.e., if they maintain the count or not).
In practice in won't matter from a performance standpoint.